test/extended/cli/explain.go

```diff
@@ -416,27 +447,47 @@ var (
 			pattern: `DESCRIPTION\:.*`,
 		},
 	}
-)
 
-func getCrdTypes(oc *exutil.CLI) []schema.GroupVersionResource {
-	configClient := configclient.NewForConfigOrDie(oc.AdminConfig())
-	clusterVersion, err := configClient.ConfigV1().ClusterVersions().Get(context.TODO(), "version", metav1.GetOptions{})
-	if err != nil {
-		e2e.Failf("Failed to get cluster version: %v", err)
+	microShiftTypes = []explainExceptions{
+		{
+			gv:      schema.GroupVersion{Group: "authorization.openshift.io", Version: "v1"},
+			field:   "rolebindingrestrictions.spec",
+			pattern: `FIELDS\:.*`,
+		},
+		{
+			gv:      schema.GroupVersion{Group: "security.internal.openshift.io", Version: "v1"},
+			field:   "rangeallocations.range",
+			pattern: `FIELD\: +range`,
+		},
+		{
+			gv:      schema.GroupVersion{Group: "security.openshift.io", Version: "v1"},
+			field:   "securitycontextconstraints",
+			pattern: `FIELDS\:.*`,
+		},
 	}
+)
 
+func getCrdTypes(oc *exutil.CLI) []schema.GroupVersionResource {
```

Comment:
`getCrdTypes` is invoked by two tests:
- "list uncovered GroupVersionResources"
- "should contain proper spec+status for CRDs [apigroup:config.openshift.io]"

The second already assumes existence of the config apigroup which is getting detected inside `getCrdTypes`.

`getCrdTypes` is always constructing the list of crd types from `baseCRDTypes`, `autoscalingTypes`, `machineTypes` and `additionalOperatorTypes` which belong to their corresponding groups as well. Which at least requires to label both tests with the corresponding groups as well. Making both tests as another place for refactoring and looping through through the apigroups.

@pacevedom this tests requires more refactoring than anticipated. Though, better to do it properly now than later/never. Nevertheless, very good progress here.

Reply:
"should contain proper spec+status for CRDs [apigroup:config.openshift.io]" requires config API because it checks for control plane topology in order to select different CRDs.
The nature of this test is to check explicitly for presence of those APIs that get built in `getCRDTypes`. If they are not present (or are poorly defined) then that should be an error. Maybe we should have a MicroShift specific test with the CRDs we install there instead? To me it is not only about checking that whatever is present is ok, it is also about specifying what should be installed.

"list uncovered GroupVersionResources" goes after API groups not covered in this file to list them. It does not perform any checks besides producing logs for those uncovered.

Reply:
`should contain proper spec+status for CRDs [apigroup:config.openshift.io]` checks presence of explanation text for corresponding CRD's spec and status.

> The nature of this test is to check explicitly for presence of those APIs that get built in getCRDTypes. If they are not present (or are poorly defined) then that should be an error.

The test does not check whether a cluster has a required list of CRDs present.  Unless I am missing the specific line.
The test checks whether CRDs listed by `getCRDTypes` have the spec and status properly described when running `oc explain` command. I.e. testing implication (is a given crd present -> has the crd spec/status properly explained). For that the list of CRDs can still be broken down based on the apigroup and each apigroup test separately. If there's a need to explicitly check for a presence of a specific list of CRDs, new test needs to be written.

> To me it is not only about checking that whatever is present is ok, it is also about specifying what should be installed.

I don't think this is the case here.

> "list uncovered GroupVersionResources" goes after API groups not covered in this file to list them. It does not perform any checks besides producing logs for those uncovered.

Ack. Reading the code a second time `resourceMap` gets populated from what's installed on a cluster minutes what's provided by `getCrdTypes`. Taking my comment back. Thanks for the pointers.